README
NAME

    Passwd::Keyring::KDEWallet - Password storage implementation based on
    KDE Wallet.

VERSION

    Version 0.2008

SYNOPSIS

    KDE Wallet based implementation of Passwd::Keyring.

        use Passwd::Keyring::KDEWallet;
    
        my $keyring = Passwd::Keyring::KDEWallet->new(
             app=>"blahblah scraper",
             group=>"Johnny web scrapers",
        );
    
        my $username = "John";  # or get from .ini, or from .argv...
    
        my $password = $keyring->get_password($username, "blahblah.com");
        unless( $password ) {
            $password = <somehow interactively prompt for password>;
    
            # securely save password for future use
            $keyring->set_password($username, $password, "blahblah.com");
        }
    
        login_somewhere_using($username, $password);
        if( password_was_wrong ) {
            $keyring->clear_password($username, "blahblah.com");
        }

    Note: see Passwd::Keyring::Auto::KeyringAPI for detailed comments on
    keyring method semantics (this document is installed with
    Passwd::Keyring::Auto package).

SUBROUTINES/METHODS

 new(app=>'app name', group=>'passwords folder')

    Initializes the processing. Croaks if kwallet (or d-bus, or anything
    needed) does not seem to be available.

    Handled named parameters:

    - app - symbolic application name (used in "Application .... is asking
    to open the wallet" KDE Wallet prompt)

    - group - name for the password group (used as KDE Wallet folder name)

 set_password(username, password, realm)

    Sets (stores) password identified by given realm for given user

 get_password($user_name, $realm)

    Reads previously stored password for given user in given app. If such
    password can not be found, returns undef.

 clear_password($user_name, $realm)

    Removes given password (if present)

 is_persistent

    Returns info, whether this keyring actually saves passwords
    persistently.

    (true in this case)

AUTHOR

    Marcin Kasperski

    Approach inspired by http://www.perlmonks.org/?node_id=869620.

BUGS

    Please report any bugs or feature requests to issue tracker at
    https://bitbucket.org/Mekk/perl-keyring-kdewallet.

SUPPORT

    You can find documentation for this module with the perldoc command.

        perldoc Passwd::Keyring::KDEWallet

    You can also look for information at:

    http://search.cpan.org/~mekk/Passwd-Keyring-KDEWallet/

    Source code is tracked at:

    https://bitbucket.org/Mekk/perl-keyring-kdewallet

LICENSE AND COPYRIGHT

    Copyright 2012 Marcin Kasperski.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Passwd::Keyring::KDEWallet

0.2008	2015-01-06
	Adding kwalitee tests

0.2007	2013-07-04
	Distribute example script.

	(bitbucket #1) Documentation contained wrong example of using set_password

0.2006	2012-11-20
	Tests don't fail but are skipped when module is built on
	system with DBUS but without KDEWallet.

        Unifying exception texts.


0.2005	2012-11-15
	Fixed bugs in package dependencies and metadata (wrong homepage
        and repository page links etc)

0.2004	2012-11-11
	Added example

0.2003	2012-11-11
	Skipping tests if not running under desktop session, so the
        module can be installed from root shell etc.

0.2002	2012-11-09
        Documentation updates.

0.2001  2012-11-08
        Small documentation fixes, more info in CPAN metadata

0.20    2012-11-08
        First version (working, functionality equivalent to Passwd::Keyring::Gnome)



Build.PL
use 5.006;
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'Passwd::Keyring::KDEWallet',
    license             => 'perl',
    dynamic_config      => 1,

    build_requires => {
        'Module::Build' => '0.19', # xs
        'Test::More' => 0,
        'Test::Pod::Coverage' => '1.0',
        'Pod::Readme' => 0,
    },
    configure_requires => {
        'Module::Build' => '0.36',
    },
    requires => {
        'perl' => 5.006,
        'Net::DBus' => 0,
    },

    share_dir => {
        dist => [ 'example' ],
    },

    create_makefile_pl => 'traditional',
    add_to_cleanup      => [ 'Passwd-Keyring-KDEWallet-*' ],

    create_readme       => 1,
    dist_author         => q{Marcin Kasperski <Marcin.Kasperski@mekk.waw.pl>},
    dist_version_from   => 'lib/Passwd/Keyring/KDEWallet.pm',
    meta_merge => {
        keywords => [ qw/ passwords security secure-storage keyring KDE KDEWallet / ],
        resources => {
            license     => 'http://dev.perl.org/licenses/',
            homepage    => 'http://bitbucket.org/Mekk/perl-keyring-kdewallet',
            repository => 'http://bitbucket.org/Mekk/perl-keyring-kdewallet',
            bugtracker => 'https://bitbucket.org/Mekk/perl-keyring-kdewallet/issues',
        },
    },
);

$builder->create_build_script();


example/imap_query_example.pl
#!perl
#
# Simple keyring usage example: check IMAP folder for newest emails.
# Non-sensitive data are taken from program args, but password is 
# got directly from the user and saved in keyring.
#
# Try running app a few times. Only on first run there should be
# password prompt.
#
# Example usage (for gmail):
#
# perl imap_query_kdewallet_sample.pl --machine imap.gmail.com --user joe2774 

{
    package RunMe;
    use Moose;
    with 'MooseX::Getopt';

    use Passwd::Keyring::KDEWallet; # Keyring itself
    use Term::ReadKey;              # For secure password prompt
    use Net::IMAP::Simple::SSL;     # We access IMAP as illustration
    use Email::Simple;              # and get some email data.

    has 'machine' => (
        is=>'ro', isa=>'Str', required=>1,
        documentation=>'IMAP machine name, like imap.gmail.com');
    has 'port' =>  (is=>'ro', isa=>'Int', required=>1, default=> 993,
                   documentation=>'IMAP port (default 993)');
    has 'user' => (is=>'ro', isa=>'Str', required=>1,
                   documentation=>'Your username on IMAP server');
    has 'folder' => (is=>'ro', isa=>'Str', required=>1, default=>'INBOX',
                     documentation=>'The folder to check');
    has 'count' => (is=>'ro', isa=>'Int', required=>1, default=> 5,
                    documentation=>'Number of emails printed');

    my $GROUP = 'Passwd::Keyring tests and samples';
    my $APP = 'imap_query_kdewallet_sample';

    my $ATTEMPTS_COUNT = 3;

    sub run {
        my $self = shift;

        my $keyring = Passwd::Keyring::KDEWallet->new(
            app=>$APP, group=>$GROUP);

        my $imap_addr = $self->machine . ':' . $self->port;
        my $imap = Net::IMAP::Simple::SSL->new($imap_addr)
          or die "Can't connect to $imap_addr: $Net::IMAP::Simple::errstr\n";

        my $realm = "IMAP:$imap_addr";
        my $user = $self->user;
        # Attempt to recover previously saved password
        my $password = $keyring->get_password($user, $realm);
        for (my $attempt_no = 1; 1; ++ $attempt_no) {
            unless($password) {
                print "Enter password for $user on $realm: ";
                ReadMode 'noecho';
                $password = ReadLine 0; chomp($password);
                ReadMode 'normal';
                print "\n";
            }
            if( $imap->login($user, $password) ) {
                # Saving the password for future
                $keyring->set_password($user, $password, $realm);
                last; # Password OK, continuing work
            } else {
                print "IMAP login failed (bad password?). Error: " . $imap->errstr, "\n";
                # Clearing the bad password in case it was taken from keyring
                $keyring->clear_password($self->user, $realm);
                $password = '';
                # Retrying unless we exhaused attempts
                if($attempt_no >= $ATTEMPTS_COUNT) {
                    die "$attempt_no login failures, good bye\n";
                }
            }
        }

        # Actual work, as illustration

        my $folder = $self->folder;
        my $count = $imap->select($folder);
        unless(defined($count)) {
            die "Unable to open folder $folder (maybe bad name?): $Net::IMAP::Simple::errstr\n";
        }
        print "$count messages in $folder. Newest:\n";
        for(my $idx = $count;
            $idx > 0 && $idx > $count - $self->count;
            -- $idx) {
            my $email = Email::Simple->new(join '', @{ $imap->top($idx) } );
            printf("[%03d] %s\n", $idx, $email->header('Subject'));
        }
    }
}

my $run_me = RunMe->new_with_options();
$run_me->run;


lib/Passwd/Keyring/KDEWallet.pm
package Passwd::Keyring::KDEWallet;

use warnings;
use strict;
#use parent 'Keyring';
use Net::DBus;
use Carp;

=head1 NAME

Passwd::Keyring::KDEWallet - Password storage implementation based on KDE Wallet.

=head1 VERSION

Version 0.2008

=cut

our $VERSION = '0.2008';

our $APP_NAME = "Passwd::Keyring";
our $FOLDER_NAME = "Perl-Passwd-Keyring";

=head1 SYNOPSIS

KDE Wallet based implementation of L<Passwd::Keyring>.

    use Passwd::Keyring::KDEWallet;

    my $keyring = Passwd::Keyring::KDEWallet->new(
         app=>"blahblah scraper",
         group=>"Johnny web scrapers",
    );

    my $username = "John";  # or get from .ini, or from .argv...

    my $password = $keyring->get_password($username, "blahblah.com");
    unless( $password ) {
        $password = <somehow interactively prompt for password>;

        # securely save password for future use
        $keyring->set_password($username, $password, "blahblah.com");
    }

    login_somewhere_using($username, $password);
    if( password_was_wrong ) {
        $keyring->clear_password($username, "blahblah.com");
    }

Note: see L<Passwd::Keyring::Auto::KeyringAPI> for detailed comments
on keyring method semantics (this document is installed with
C<Passwd::Keyring::Auto> package).

=head1 SUBROUTINES/METHODS

=head2 new(app=>'app name', group=>'passwords folder')

Initializes the processing. Croaks if kwallet (or d-bus, or anything needed) does not 
seem to be available.

Handled named parameters: 

- app - symbolic application name (used in "Application .... is asking to open the wallet"
  KDE Wallet prompt)

- group - name for the password group (used as KDE Wallet folder name)

=cut

sub new {
    my ($cls, %args) = @_;

    my $self = {};
    $self->{app} = $args{app} || 'Passwd::Keyring::KDEWallet';
    $self->{group} = $args{group} || 'Passwd::Keyring::default';
    bless $self;

    #$self->{bus} = Net::DBus->find()
    $self->{bus} = Net::DBus->session()
      or croak("KWallet not available (can't access DBus)");
    # get_service also may fail by itself, I got cpantesters reports with message
    # "org.freedesktop.DBus.Error.ServiceUnknown: The name org.kde.kwalletd was not provided by any .service files"
    # Let's rewrite this slightly
    my$kwallet_svc;
    eval {
        $kwallet_svc = $self->{bus}->get_service('org.kde.kwalletd');
    };
    if($@) {
        croak("KWallet not available (not installed?). Details:\n$@");
    } elsif (! $kwallet_svc) {
        croak("KWallet not available (can't access KWallet, likely kwalletd not running)");
    }
    $self->{kwallet} = $kwallet_svc->get_object('/modules/kwalletd', 'org.kde.KWallet')
      or croak("Kwallet not available (can't find wallet)");
    $self->_open_if_not_open();

    unless($self->{kwallet}->hasFolder($self->{handle}, $self->{group}, $self->{app})) {
        $self->{kwallet}->createFolder($self->{handle}, $self->{group}, $self->{app})
          or croak("Failed to create $self->{group} folder (app $self->{app})");
    }

    return $self;
}

sub _open_if_not_open {
    my $self = shift;

    if($self->{handle}) {
        if($self->{kwallet}->isOpen($self->{handle})) {
            return;
        }
    }
    my $net_wallet = $self->{kwallet}->networkWallet()
      or croak("Kwallet not available (can't access network wallet");
    $self->{handle} = $self->{kwallet}->open($net_wallet, 0, $self->{app})
      or croak("Failed to open the KDE wallet");
}

=head2 set_password(username, password, realm)

Sets (stores) password identified by given realm for given user 

=cut

sub set_password {
    my ($self, $user_name, $user_password, $realm) = @_;
    $self->_open_if_not_open();
    my $status = $self->{kwallet}->writePassword(
        $self->{handle}, $self->{group}, "$realm || $user_name", $user_password, $self->{app});
    if($status) { # non-zero means failure
        croak("Failed to save the password (status $status, user name $user_name, realm $realm, handle $self->{handle}, group $self->{group})");
    }
}

=head2 get_password($user_name, $realm)

Reads previously stored password for given user in given app.
If such password can not be found, returns undef.

=cut

sub get_password {
    my ($self, $user_name, $realm) = @_;
    $self->_open_if_not_open();
    my $reply = $self->{kwallet}->readPassword(
        $self->{handle}, $self->{group}, "$realm || $user_name", $self->{app});
    # In case of missing passsword we get empty string. I do not know
    # whether it is possible to distinguish missing password from empty password,
    # but empty passwords are exotic enough to ignore.
    return undef if ! defined($reply) or $reply eq '';
    return $reply;
}

=head2 clear_password($user_name, $realm)

Removes given password (if present)

=cut

sub clear_password {
    my ($self, $user_name, $realm) = @_;
    $self->_open_if_not_open();
    my $status = $self->{kwallet}->removeEntry(
        $self->{handle}, $self->{group}, "$realm || $user_name", $self->{app});
    if($status == 0) {
        return 1;
    } else {
        # TODO: classify failures
        return 0;
    }
}

=head2 is_persistent

Returns info, whether this keyring actually saves passwords persistently.

(true in this case)

=cut

sub is_persistent {
    my ($self) = @_;
    return 1;
}

=head1 AUTHOR

Marcin Kasperski

Approach inspired by L<http://www.perlmonks.org/?node_id=869620>.

=head1 BUGS

Please report any bugs or feature requests to 
issue tracker at L<https://bitbucket.org/Mekk/perl-keyring-kdewallet>.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Passwd::Keyring::KDEWallet

You can also look for information at:

L<http://search.cpan.org/~mekk/Passwd-Keyring-KDEWallet/>

Source code is tracked at:

L<https://bitbucket.org/Mekk/perl-keyring-kdewallet>

=head1 LICENSE AND COPYRIGHT

Copyright 2012 Marcin Kasperski.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut


1; # End of Passwd::Keyring::KDEWallet


